drivers/periph/spi: added means to make SPI thread-safe

haukepetersen opened the pull request:
extended the SPI driver's API with locking/unlocking functions and implemented them exemplary for the `stm32f4`.

These functions are needed for applications that share a SPI bus for more then one slave device, e.g. 2 sensors, while their drivers run in separate threads.

Usage:
```c
...
spi_aquire(SPI_x);
gpio_set(CS_PIN);
spi_transfer(SPI_X, ...);
gpio_clear(CS_PIN);
spi_release(SPI_X);
...
```

The price to pay:
```
for 1 configured SPI interface:    84 byte ROM and  8 byte RAM
for 2 configured SPI interfaces:  112 byte ROM and 16 byte RAM
(compiling the tests/periph_spi application)
```

gebart commented:
I think the memory bytes is a small price to pay for the thread safety of our device drivers. Errors caused by these kinds of race conditions will be extremely difficult to reproduce or locate when they bite you, adding a simple NOP somewhere completely unrelated could potentially make such a problem hide again.

:+1: for the API, it is similar to what is being used in ChibiOS (which has been around for a very long time), so the general idea has already been tested thoroughly.

I was thinking of whether to put `spi_acquire` in a common file for all ports, but I believe that it is better to keep all SPI related functions in one file. It also opens the possibility for some corner cases where you would want to customize the locking mechanism (I have something in mind for `kinetis_common` #2265 ).

PeterKietzmann commented:
I also think that the price is ok for the benefit we get. Should we maybe adapt the test? Or extend it?

haukepetersen commented:
@gebart I wouldn't put the locking in a central file, as there might be platforms where the locking can be handled more efficient without mutexes or similar. Also the additional implementation effort is not too high...

@PeterKietzmann I don't think we need to adapt the test, the Mutexes itself are well tested and I think the effort for a 'good' test are not worth it (as the locking implementation is quite straight forward...).

PeterKietzmann commented:
Well, I would ACK this. Any other opinions?

BytesGalore commented:
looks also good from my side

thomaseichinger commented on drivers/include/periph/spi.h:
```diff
@@ -123,6 +123,28 @@ int spi_init_slave(spi_t dev, spi_conf_t conf, char (*cb)(char data));
 int spi_conf_pins(spi_t dev);
 
 /**
+ * @brief Get mutual access to the given SPI bus
```
Maybe mutual **exclusive** access?

gebart replied:
..."mutual**ly** exclusive"...

thomaseichinger commented on cpu/stm32f4/periph/spi.c:
```diff
@@ -102,6 +111,9 @@ int spi_init_master(spi_t dev, spi_conf_t conf, spi_speed_t speed)
     }
 
+    /* initialize lock */
+    mutex_init(&spi_config[dev].lock);
```
Maybe move this to the top of the init function followed by a `spi_acquire`?

gebart replied:
Does mutex_init() handle multiple initializations? It would be useful to prevent multiple calls to spi_init from trampling each other.
On the other hand, I would prefer that all calls to `spi_init` would be moved to each platform's `board.c` instead of letting the device driver reinitialize the SPI bus after another device driver has already began using it. For example: at86rf231 contains a call to `spi_init` inside its init function.

BytesGalore replied:
@gebart  basically the mutex initialization [resets](https://github.com/RIOT-OS/RIOT/blob/master/core/include/mutex.h#L61) the mutex. 
In this case initializing, or resetting, a specific `dev` into SPI master mode must reset the mutex too as it could be persist locked from a previous state and produce a deadlock.  

> On the other hand, I would prefer that all calls to spi_init would be moved to each platform's board.c instead of...

Another proposition to prevent reinitialization of the mutex and a running device would be to store the initialized state in the `spi_state_t` of the `dev` and to skip `spi_init_master()` for the `dev` if its done already.
Resetting the stored state could be done in [`spi_poweroff()`](https://github.com/haukepetersen/RIOT/blob/opt_spi_threadsafe/cpu/stm32f4/periph/spi.c#L429) for the specific `dev`.

+1 to move the mutex to the top and acquire it before doing initializations

gebart replied:
OT: The problem with multiple calls to spi_init is that they can mess with each others' settings, e.g. baud rate. If there are two devices on the bus with different maximum speeds the code needs to be configured for the lowest baud rate of the two unless we want to manually switch rate between each SPI transfer.

PeterKietzmann replied:
@gebart, wouldn't the porting of `spi_init`-calls to `board.c` lead to the same deficit?

gebart replied:
@PeterKietzmann Moving it to board.c (and a common configuration in board.h or periph_conf.h) would be the logical place to have it since when you are designing the hardware and application you will have to decide on a common subset of the available settings: we have to obey the requirements of the slowest device, we need to find a common clock polarity configuration that works on all of the connected devices etc.

haukepetersen commented:
here is an update - the mutex initialization issue is fixed through static initialization - multiple calls to `spi_init` should now be save.

I decided against calling `spi_acquire` in the init function. If you application uses one SPI bus for more then one device (with differing configurations), I think it should be up to the user to protect the complete acquire->init->transfer->release sequence. This is possible as the mutex works already before init is called the first time...

haukepetersen commented:
an, and the new change even saved even 24 bytes in ROM :-)

thomaseichinger commented:
Looks good to me.

PeterKietzmann commented:
I am fine with this soultion and willing to ACK. But also I would like to hear @gebart 's optinion on it.

gebart commented:
@haukepetersen While it is still most likely a user error, I think it would be beneficial for developers to have the init function lock the mutex while initializing, to avoid any race conditions between two drivers, or between an application main thread and some data consumer thread. I don't really see why we should _not_ lock the mutex for all of `spi_init`?

haukepetersen commented:
@gebart You are right, the overhead is limited and it could save some trouble. I was first reluctant against this idea since the major design goal was to keep the peripheral drivers as slim as possible - and every function call adds some more bytes to them... But I agree, the benefits for locking the init function are greater then the drawbacks of 2 function calls, so I will update this PR as soon as I have a minute.

gebart commented:
@haukepetersen I will update #2317 accordingly.

gebart commented:
@PeterKietzmann I think this is a good solution to the problem. The locking inside `spi_init` should be added before merging though.

thomaseichinger commented:
As the mutexes (mutices?) are statically initialised I thought usage would be
```C
spi_acquire(dev);
spi_init_master(dev, ...);
spi_release(dev);
```
Wouldn't this result in a more consistent usage?

gebart commented:
@thomaseichinger It would make the API usage more consistent, yes. I am updating #2317 again.

haukepetersen commented:
I am completely for @thomaseichinger's solution! I think it is for (i) more consitent and it (ii) leaves it to the user to omit the locking for setups where I can sure that I am only using one SPI device for example...

squashed and rebased - ready for merge?

gebart commented:
ACK

haukepetersen commented:
ok, somehow Travis is complaining that the PR needs to be squashed, though the label was removed... Do we have a bug in our Travis script?

PeterKietzmann commented:
Travis is happy

haukepetersen commented:
finally. go then!
